File: src/roulette/roulette.ts
import { Lottery } from '../games/lottery'

import type { AppContext } from '../di'
import type { RouletteLottery } from './store'
import type { GuildMember, Interaction } from '../discord/types'
import { bgrLabel } from '../discord/api'

export const rouletteTimeSeconds = 30
// export const rouletteTimeSeconds = 5 // debug
export const rouletteTimeMs = rouletteTimeSeconds * 1000
// const updateInterval =

export interface RouletteProps {
  context: AppContext
}

export interface NewLotteryProps {
  creator: GuildMember
  bet: number
}

export class Roulette {
  public readonly lottery: RouletteLottery
  private readonly context: AppContext

  static init(props: RouletteProps & NewLotteryProps): { roulette?: Roulette; error?: Error } {
    try {
      const roulette = new Roulette({
        lottery: new Lottery({ ...props, players: [props.creator] }),
        ...props,
      })
      return { roulette }
    } catch (error) {
      return { error }
    }
  }

  static async load(
    id: string,
    props: RouletteProps
  ): Promise<{ roulette?: Roulette; error?: Error }> {
    try {
      const lottery = await props.context.rouletteStore.get(id)
      if (!lottery) throw new Error('Roulette Lottery not found')
      return { roulette: new Roulette({ lottery, ...props }) }
    } catch (error) {
      return { error }
    }
  }

  private constructor({ context, lottery }: RouletteProps & { lottery: RouletteLottery }) {
    this.lottery = lottery
    this.context = context
  }

  async start(interaction: Interaction): Promise<Date> {
    this.lottery.start()

    await this.context.rouletteStore.put(this.lottery)

    await this.context.jobQueue.enqueue(
      'roulette:finish',
      { id: this.lottery.id, interaction },
      rouletteTimeSeconds
    )

    // We called start() above, this can't be empty
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this.lottery.startTime!
  }

  getStart(): Date | undefined {
    return this.lottery.startTime
  }

  get id(): string {
    return this.lottery.id
  }

  get buyIn(): number {
    return this.lottery.buyIn
  }

  getBet(): number {
    return this.lottery.bet
  }

  getCreator(): GuildMember {
    return this.lottery.creator
  }

  getPlayers(): GuildMember[] {
    return this.lottery.players
  }

  async addPlayer(player: GuildMember): Promise<void> {
    this.lottery.addPlayer(player)
    await this.context.rouletteStore.setPlayers(this.lottery.id, this.lottery.players)
  }

  async finish(): Promise<string> {
    if (!this.lottery.canFinish()) {
      return `${this.lottery.creator.username}'s roulette game was cancelled, not enough players joined.`
    }

    const { winner, payouts } = this.lottery.finish()
    const names = payouts.map(([player]) => player.username)

    await this.context.userStore.incrementUserReps(
      ...payouts.map(([member, offset]) => ({ member, offset }))
    )

    await this.context.rouletteStore.delete(this.lottery.id)

    // TODO get all new rep values and include in message
    return `The roulette game has ended. ${names.join(', ')} all bet ${bgrLabel(this.getBet())}. ${
      winner.username
    } won ${bgrLabel(this.lottery.potSize)}`
  }
}

File: src/roulette/roulette.rs
use firestore::FirestoreDb;
use serde::{Deserialize, Serialize};

use crate::discord::helpers::bgr_label;
use crate::discord::types::GuildMember;
use crate::games::lottery::{Lottery, StoredPlayer};
use crate::jobs::JobType;
use crate::roulette::store::{RouletteStore, RouletteLottery};
use crate::users::UserStore;

pub const ROULETTE_TIME_SECONDS: u64 = 30;
pub const ROULETTE_TIME_MS: u64 = ROULETTE_TIME_SECONDS * 1000;

#[derive(Debug, Serialize, Deserialize)]
pub struct RouletteJobPayload {
    pub id: String,
    pub interaction_token: String,
}

pub struct Roulette {
    pub lottery: RouletteLottery,
    store: RouletteStore,
}

impl Roulette {
    pub fn init(
        db: &FirestoreDb,
        creator: &GuildMember,
        bet: i64,
    ) -> anyhow::Result<Self> {
        let stored_creator = StoredPlayer::from(creator);
        let mut lottery = Lottery::new(stored_creator.clone(), bet)?;
        lottery.add_player(stored_creator);
        Ok(Self {
            lottery,
            store: RouletteStore::new(db.clone()),
        })
    }

    pub async fn load(db: &FirestoreDb, id: &str) -> anyhow::Result<Self> {
        let store = RouletteStore::new(db.clone());
        let lottery = store
            .get(id)
            .await?
            .ok_or_else(|| anyhow::anyhow!("Roulette lottery not found"))?;
        Ok(Self { lottery, store })
    }

    pub async fn start(&mut self, interaction_token: &str, job_queue: &crate::jobs::JobQueue) -> anyhow::Result<String> {
        let start_time = self.lottery.start();

        self.store.put(&self.lottery).await?;

        let payload = RouletteJobPayload {
            id: self.lottery.id.clone(),
            interaction_token: interaction_token.to_string(),
        };

        job_queue
            .enqueue(
                JobType::RouletteFinish,
                serde_json::to_value(&payload)?,
                ROULETTE_TIME_SECONDS,
            )
            .await?;

        Ok(start_time)
    }

    pub fn id(&self) -> &str {
        &self.lottery.id
    }

    pub fn buy_in(&self) -> i64 {
        self.lottery.buy_in()
    }

    pub fn bet(&self) -> i64 {
        self.lottery.bet
    }

    pub fn creator(&self) -> &StoredPlayer {
        &self.lottery.creator
    }

    pub fn players(&self) -> &[StoredPlayer] {
        &self.lottery.players
    }

    pub fn start_time(&self) -> Option<&String> {
        self.lottery.start_time.as_ref()
    }

    pub async fn add_player(&mut self, player: &GuildMember) -> anyhow::Result<()> {
        let stored = StoredPlayer::from(player);
        self.lottery.add_player(stored);
        self.store
            .set_players(&self.lottery.id, &self.lottery.players)
            .await?;
        Ok(())
    }

    pub async fn finish(&self, user_store: &UserStore) -> anyhow::Result<String> {
        if !self.lottery.can_finish() {
            self.store.delete(&self.lottery.id).await?;
            return Ok(format!(
                "{}'s roulette game was cancelled, not enough players joined.",
                self.lottery.creator.username
            ));
        }

        let result = self.lottery.finish();
        let names: Vec<&str> = result.payouts.iter().map(|(p, _)| p.username.as_str()).collect();

        // Convert payouts to GuildMember + offset pairs for user store
        let updates: Vec<(GuildMember, i64)> = result
            .payouts
            .iter()
            .map(|(p, offset)| (GuildMember::from(p), *offset))
            .collect();

        user_store.increment_user_reps(&updates).await?;

        self.store.delete(&self.lottery.id).await?;

        Ok(format!(
            "The roulette game has ended. {} all bet {}. {} won {}",
            names.join(", "),
            bgr_label(self.bet(), false),
            result.winner.username,
            bgr_label(self.lottery.pot_size(), false),
        ))
    }
}
